File: mifosng-android/src/main/java/com/mifos/mifosxdroid/dialogfragments/chargedialog/ChargeDialogFragment.java
/*
 * This project is licensed under the open source MPL V2.
 * See https://github.com/openMF/android-client/blob/master/LICENSE.md
 */

package com.mifos.mifosxdroid.dialogfragments.chargedialog;

import android.app.Dialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;

import com.google.android.material.textfield.MaterialAutoCompleteTextView;
import com.mifos.mifosxdroid.R;
import com.mifos.mifosxdroid.core.MifosBaseActivity;
import com.mifos.mifosxdroid.core.ProgressableDialogFragment;
import com.mifos.mifosxdroid.core.util.Toaster;
import com.mifos.mifosxdroid.uihelpers.MFDatePicker;
import com.mifos.objects.client.ChargeCreationResponse;
import com.mifos.objects.client.Charges;
import com.mifos.objects.templates.clients.ChargeTemplate;
import com.mifos.services.data.ChargesPayload;
import com.mifos.utils.Constants;
import com.mifos.utils.DateHelper;
import com.mifos.utils.FragmentConstants;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by nellyk on 1/22/2016.
 * <p/>
 * Use this Dialog Fragment to Create and/or Update charges
 */
public class ChargeDialogFragment extends ProgressableDialogFragment implements
        MFDatePicker.OnDatePickListener, ChargeDialogMvpView  {

    public final String LOG_TAG = getClass().getSimpleName();

    @BindView(R.id.chargeNameField)
    MaterialAutoCompleteTextView spChargeName;

    @BindView(R.id.amount_due_charge)
    EditText etAmountDue;

    @BindView(R.id.et_date)
    EditText etChargeDueDate;

    @BindView(R.id.et_charge_locale)
    EditText etChargeLocale;

    @BindView(R.id.bt_save_charge)
    Button btnSaveCharge;

    @Inject
    ChargeDialogPresenter mChargeDialogPresenter;

    @Nullable
    private OnChargeCreateListener chargeCreateListener;

    private List<String> chargeNameList = new ArrayList<>();
    private ArrayAdapter<String> chargeNameAdapter;
    private ChargeTemplate mChargeTemplate;
    private String dueDateString;
    private List<Integer> dueDateAsIntegerList;
    private View rootView;
    private DialogFragment mfDatePicker;
    private int chargeId;
    private String chargeName;
    private int clientId;
    private Charges createdCharge;

    public static ChargeDialogFragment newInstance(int clientId) {
        ChargeDialogFragment chargeDialogFragment = new ChargeDialogFragment();
        Bundle args = new Bundle();
        args.putInt(Constants.CLIENT_ID, clientId);
        chargeDialogFragment.setArguments(args);
        return chargeDialogFragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((MifosBaseActivity) getActivity()).getActivityComponent().inject(this);
        if (getArguments() != null)
            clientId = getArguments().getInt(Constants.CLIENT_ID);
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {

        if (getArguments() != null)
            clientId = getArguments().getInt(Constants.CLIENT_ID);
        return super.onCreateDialog(savedInstanceState);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if(getDialog() != null && getDialog().getWindow() != null)
            getDialog().getWindow().setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle
            savedInstanceState) {

        // Inflate the layout for this fragment
        if (getActivity().getActionBar() != null)
            getActivity().getActionBar().setDisplayHomeAsUpEnabled(true);
        rootView = inflater.inflate(R.layout.dialog_fragment_charge, null);

        ButterKnife.bind(this, rootView);
        mChargeDialogPresenter.attachView(this);
        inflatedueDate();
        inflateChargesSpinner();
        inflateChargeNameSpinner();

        return rootView;
    }

    @OnClick(R.id.bt_save_charge)
    public void saveNewCharge() {
        //Insert values for the new Charge.
        if (etAmountDue.getText().toString().isEmpty()) {
            Toaster.show(rootView, getString(R.string.amount)
                    + " " + getString(R.string.error_cannot_be_empty));
            return;
        }
        createdCharge = new Charges();
        createdCharge.setChargeId(chargeId);
        createdCharge.setAmount(Double.parseDouble(etAmountDue.getEditableText().toString()));

        dueDateAsIntegerList = DateHelper.convertDateAsReverseInteger(dueDateString);
        createdCharge.setDueDate(dueDateAsIntegerList);
        createdCharge.setName(chargeName);

        ChargesPayload chargesPayload = new ChargesPayload();
        chargesPayload.setAmount(etAmountDue.getEditableText().toString());
        chargesPayload.setLocale(etChargeLocale.getEditableText().toString());
        chargesPayload.setDueDate(dueDateString);
        chargesPayload.setDateFormat("dd MMMM yyyy");
        chargesPayload.setChargeId(chargeId);
        initiateChargesCreation(chargesPayload);
    }

    @Override
    public void onDatePicked(String date) {
        dueDateString = DateHelper.getDateAsStringUsedForCollectionSheetPayload(date)
                .replace("-", " ");
        etChargeDueDate.setText(dueDateString);
    }

    //Charges Fetching API
    private void inflateChargesSpinner() {
        mChargeDialogPresenter.loadAllChargesV2(clientId);
    }

    //Charges Creation APi
    private void initiateChargesCreation(ChargesPayload chargesPayload) {
        mChargeDialogPresenter.createCharges(clientId, chargesPayload);
    }

    public void inflatedueDate() {
        mfDatePicker = MFDatePicker.newInsance(this);
        String receivedDate = MFDatePicker.getDatePickedAsString();
        dueDateString = DateHelper.getDateAsStringUsedForCollectionSheetPayload(receivedDate)
                .replace("-", " ");
        dueDateAsIntegerList = DateHelper.convertDateAsListOfInteger(dueDateString);
        etChargeDueDate.setText(dueDateString);
    }

    @OnClick(R.id.et_date)
    public void inflateDatePicker() {
        mfDatePicker.show(getActivity().getSupportFragmentManager(), FragmentConstants
                .DFRAG_DATE_PICKER);
    }


    public void inflateChargeNameSpinner() {
        chargeNameAdapter = new ArrayAdapter<>(getActivity(),
                android.R.layout.simple_spinner_item, chargeNameList);
        chargeNameAdapter
                .setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spChargeName.setAdapter(chargeNameAdapter);
        spChargeName.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int position, long l) {
                chargeId = mChargeTemplate.getChargeOptions().get(position).getId();
                chargeName = mChargeTemplate.getChargeOptions().get(position).getName();
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });
    }

    @Override
    public void showAllChargesV2(ChargeTemplate chargeTemplate) {
        mChargeTemplate = chargeTemplate;
        chargeNameList.addAll(mChargeDialogPresenter.filterChargeName
                (chargeTemplate.getChargeOptions()));
        chargeNameAdapter.notifyDataSetChanged();
    }

    @Override
    public void showChargesCreatedSuccessfully(ChargeCreationResponse chargeCreationResponse) {
        if (chargeCreateListener != null) {
            createdCharge.setClientId(chargeCreationResponse.getClientId());
            createdCharge.setId(chargeCreationResponse.getResourceId());
            chargeCreateListener.onChargeCreatedSuccess(createdCharge);
        } else {
            Toaster.show(rootView, getString(R.string.message_charge_created_success));
        }
        getDialog().dismiss();
    }

    @Override
    public void showChargeCreatedFailure(String errorMessage) {
        if (chargeCreateListener != null) {
            chargeCreateListener.onChargeCreatedFailure(errorMessage);
        } else {
            Toaster.show(rootView, errorMessage);
        }
    }

    @Override
    public void showFetchingError(String s) {
        Toaster.show(rootView, s);
    }

    @Override
    public void showProgressbar(boolean b) {
        showProgress(b);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mChargeDialogPresenter.detachView();
    }

    public void setOnChargeCreatedListener(@Nullable OnChargeCreateListener chargeCreatedListener) {
        this.chargeCreateListener = chargeCreatedListener;
    }
}

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/dialogfragments/chargedialog/ChargeDialogFragment.kt
/*
 * This project is licensed under the open source MPL V2.
 * See https://github.com/openMF/android-client/blob/master/LICENSE.md
 */
package com.mifos.mifosxdroid.dialogfragments.chargedialog

import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.ViewModelProvider
import com.mifos.core.data.ChargesPayload
import com.mifos.core.objects.client.ChargeCreationResponse
import com.mifos.core.objects.client.Charges
import com.mifos.core.objects.templates.clients.ChargeTemplate
import com.mifos.mifosxdroid.R
import com.mifos.mifosxdroid.core.ProgressableDialogFragment
import com.mifos.mifosxdroid.core.util.Toaster.show
import com.mifos.mifosxdroid.databinding.DialogFragmentChargeBinding
import com.mifos.mifosxdroid.uihelpers.MFDatePicker
import com.mifos.mifosxdroid.uihelpers.MFDatePicker.Companion.newInsance
import com.mifos.mifosxdroid.uihelpers.MFDatePicker.OnDatePickListener
import com.mifos.utils.Constants
import com.mifos.utils.DateHelper
import com.mifos.utils.FragmentConstants
import dagger.hilt.android.AndroidEntryPoint

/**
 * Created by nellyk on 1/22/2016.
 *
 *
 * Use this Dialog Fragment to Create and/or Update charges
 */
@AndroidEntryPoint
class ChargeDialogFragment : ProgressableDialogFragment(), OnDatePickListener,
    OnItemSelectedListener {
    val LOG_TAG = javaClass.simpleName

    private lateinit var binding: DialogFragmentChargeBinding

    private lateinit var viewModel: ChargeDialogViewModel

    private var chargeCreateListener: OnChargeCreateListener? = null
    private val chargeNameList: MutableList<String> = ArrayList()
    private lateinit var chargeNameAdapter: ArrayAdapter<String>
    private var mChargeTemplate: ChargeTemplate? = null
    private lateinit var dueDateString: String
    private var dueDateAsIntegerList: List<Int>? = null
    private var mfDatePicker: DialogFragment? = null
    private var chargeId = 0
    private var chargeName: String? = null
    private var clientId = 0
    private var createdCharge: Charges? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) clientId = requireArguments().getInt(Constants.CLIENT_ID)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        if (arguments != null) clientId = requireArguments().getInt(Constants.CLIENT_ID)
        return super.onCreateDialog(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {

        // Inflate the layout for this fragment
        if (requireActivity().actionBar != null) requireActivity().actionBar?.setDisplayHomeAsUpEnabled(
            true
        )
        binding = DialogFragmentChargeBinding.inflate(inflater, container, false)
        viewModel = ViewModelProvider(this)[ChargeDialogViewModel::class.java]
        inflateDueDate()
        inflateChargesSpinner()
        inflateChargeNameSpinner()

        viewModel.chargeDialogUiState.observe(viewLifecycleOwner) {
            when (it) {
                is ChargeDialogUiState.ShowAllChargesV2 -> {
                    showProgressbar(false)
                    showAllChargesV2(it.chargeTemplate)
                }

                is ChargeDialogUiState.ShowChargesCreatedSuccessfully -> {
                    showProgressbar(false)
                    showChargesCreatedSuccessfully(it.chargeCreationResponse)
                }

                is ChargeDialogUiState.ShowFetchingError -> {
                    showProgressbar(false)
                    showFetchingError(it.message)
                }

                is ChargeDialogUiState.ShowProgressbar -> showProgressbar(true)
            }
        }

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.btSaveCharge.setOnClickListener {
            saveNewCharge()
        }

        binding.etDate.setOnClickListener {
            inflateDatePicker()
        }
    }

    fun saveNewCharge() {
        //Insert values for the new Charge.
        if (binding.amountDueCharge.text.toString().isEmpty()) {
            show(
                binding.root, getString(R.string.amount)
                        + " " + getString(R.string.error_cannot_be_empty)
            )
            return
        }
        createdCharge = Charges()
        createdCharge?.chargeId = chargeId
        createdCharge?.amount = binding.amountDueCharge.editableText.toString().toDouble()
        dueDateAsIntegerList = DateHelper.convertDateAsReverseInteger(dueDateString)
        createdCharge?.dueDate = dueDateAsIntegerList as List<Int>
        createdCharge?.name = chargeName
        val chargesPayload = ChargesPayload()
        chargesPayload.amount = binding.amountDueCharge.editableText.toString()
        chargesPayload.locale = binding.etChargeLocale.editableText.toString()
        chargesPayload.dueDate = dueDateString
        chargesPayload.dateFormat = "dd MMMM yyyy"
        chargesPayload.chargeId = chargeId
        initiateChargesCreation(chargesPayload)
    }

    override fun onDatePicked(date: String?) {
        dueDateString = DateHelper.getDateAsStringUsedForCollectionSheetPayload(date)
            .replace("-", " ")
        binding.chargeDueDate.text = dueDateString
    }

    //Charges Fetching API
    private fun inflateChargesSpinner() {
        viewModel.loadAllChargesV2(clientId)
    }

    //Charges Creation APi
    private fun initiateChargesCreation(chargesPayload: ChargesPayload) {
        viewModel.createCharges(clientId, chargesPayload)
    }

    private fun inflateDueDate() {
        mfDatePicker = newInsance(this)
        val receivedDate: String? = MFDatePicker.datePickedAsString
        dueDateString = DateHelper.getDateAsStringUsedForCollectionSheetPayload(receivedDate)
            .replace("-", " ")
        dueDateAsIntegerList = DateHelper.convertDateAsListOfInteger(dueDateString)
        binding.chargeDueDate.text = dueDateString
    }

    private fun inflateDatePicker() {
        mfDatePicker?.show(
            requireActivity().supportFragmentManager,
            FragmentConstants.DFRAG_DATE_PICKER
        )
    }

    private fun inflateChargeNameSpinner() {
        chargeNameAdapter = ArrayAdapter(
            requireActivity(),
            android.R.layout.simple_spinner_item, chargeNameList
        )
        chargeNameAdapter
            .setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.spChargeName.adapter = chargeNameAdapter
        binding.spChargeName.onItemSelectedListener = this
    }

    private fun showAllChargesV2(chargeTemplate: ChargeTemplate) {
        mChargeTemplate = chargeTemplate
        viewModel.filterChargeName(chargeTemplate.chargeOptions).let { chargeNameList.addAll(it) }
        chargeNameAdapter.notifyDataSetChanged()
    }

    override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        when (parent.id) {
            R.id.sp_charge_name -> {
                chargeId = mChargeTemplate?.chargeOptions?.getOrNull(position)?.id!!
                chargeName = mChargeTemplate?.chargeOptions?.getOrNull(position)?.name
            }
        }
    }

    override fun onNothingSelected(parent: AdapterView<*>?) {}
    private fun showChargesCreatedSuccessfully(chargeCreationResponse: ChargeCreationResponse) {
        if (chargeCreateListener != null) {
            createdCharge?.clientId = chargeCreationResponse.clientId
            createdCharge?.id = chargeCreationResponse.resourceId
            createdCharge?.let { chargeCreateListener?.onChargeCreatedSuccess(it) }
        } else {
            show(binding.root, getString(R.string.message_charge_created_success))
        }
        dialog?.dismiss()
    }

    private fun showChargeCreatedFailure(errorMessage: String) {
        if (chargeCreateListener != null) {
            chargeCreateListener?.onChargeCreatedFailure(errorMessage)
        } else {
            show(binding.root, errorMessage)
        }
    }

    private fun showFetchingError(s: String) {
        show(binding.root, s)
    }

    private fun showProgressbar(b: Boolean) {
        showProgress(b)
    }

    fun setOnChargeCreatedListener(chargeCreatedListener: OnChargeCreateListener?) {
        chargeCreateListener = chargeCreatedListener
    }

    companion object {
        fun newInstance(clientId: Int): ChargeDialogFragment {
            val chargeDialogFragment = ChargeDialogFragment()
            val args = Bundle()
            args.putInt(Constants.CLIENT_ID, clientId)
            chargeDialogFragment.arguments = args
            return chargeDialogFragment
        }
    }
}